### DISCUSSION TITLE
Promise resolve not being called back within .load method

### DISCUSSION THREAD

**Discussion by filipe-costa** [pixijs/pixijs#7862] [Q&A] [answered]
Hello all!

I am currently working with Pixijs and I am using Jest to test my working code. 

I have created a [Minimum Reproducible Demo](https://github.com/filipe-costa/jest-timeout), you can run the tests for two "similar implementations" one using a regular cache and another using Pixijs for sanity checks.

I would have expected the following to work:

```typescript
return new Promise((resolve) => {
  this.loader.add(assets).load(resolve)
})
```

Therefore, since resolve() is never called Jest times out.

```
thrown: "Exceeded timeout of 5000 ms for a test.
Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."
```

What am I doing wrong or what are the best practices to achieve the same result?

Thank you in advance!

**Comment by bigtimebuddy** [comment]
Are you sure this isn't something about your jest environment? Loader does use XMLHttpRequest under the hood, are you sure that's supported in this context?

**Reply by filipe-costa** [reply]
From what I can gather through log statements, the XMLHttpRequest exists in the jest testing context. I have added two extra logs to check this, one that returns true for the existence of XHR and, another that checks that one of the resources has been loaded correctly and added to the resources cache.

Output from one of the loaded resources when logging: is the following:

```
charmander: Resource {
        _flags: 5,
        name: 'charmander',
        url: 'data:image/png;base64,...rectified due to being too long'
        extension: 'png',
        data: HTMLImageElement {},
        crossOrigin: '',
        timeout: 0,
        loadType: 2,
        xhrType: undefined,
        metadata: {},
        error: null,
        xhr: null,
        children: [],
        type: 3,
        progressChunk: 50,
        _dequeue: [Function: onceWrapper],
        _onLoadBinding: MiniSignalBinding {
          _fn: [Function: _onLoad],
          _once: true,
          _thisArg: [Loader],
          _owner: [MiniSignal],
          _prev: null,
          _next: null
        },
        _elementTimer: 0,
        _boundComplete: [Function: bound complete],
        _boundOnError: [Function: bound _onError],
        _boundOnProgress: [Function: bound _onProgress],
        _boundOnTimeout: [Function: bound _onTimeout],
        _boundXhrOnError: [Function: bound _xhrOnError],
        _boundXhrOnTimeout: [Function: bound _xhrOnTimeout],
        _boundXhrOnAbort: [Function: bound _xhrOnAbort],
        _boundXhrOnLoad: [Function: bound _xhrOnLoad],
        onStart: MiniSignal { _tail: undefined, _head: undefined },
        onProgress: MiniSignal { _tail: undefined, _head: undefined },
        onComplete: MiniSignal {
          _tail: [MiniSignalBinding],
          _head: [MiniSignalBinding]
        },
        onAfterMiddleware: MiniSignal { _tail: undefined, _head: undefined }
      }
    }
```

I would expect if the problem was with Jest, my other unit test would fail as well, please let me know your thoughts regarding this and if you have issues reproducing the same issue as well. If you can't reproduce it, I will be more than happy to raise this with Jest too!

**Reply by domis86** [reply]
ok this should give us some info:

```
charmander: Resource {
        _flags: 5,
```

Then look here: https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L505

```
    /**
     * Describes if this resource has finished loading. Is true when the resource has completely
     * loaded.
     *
     * @readonly
     * @member {boolean}
     */
    get isComplete(): boolean
    {
        return this._hasFlag(LoaderResource.STATUS_FLAGS.COMPLETE);
    }
```

STATUS_FLAGS are:

```
    export enum STATUS_FLAGS {
        /** None */
        NONE= 0,
        /** Data URL */
        DATA_URL= (1 << 0),
        /** Complete */
        COMPLETE= (1 << 1),
        /** Loading */
        LOADING= (1 << 2),
    }
```

The `_hasFlag` method is checking flags by doing "logical AND" operation on them - so `_flags: 5` means that it contains flags:
* DATA_URL (1)
* LOADING (4)

So loading is not in fact completed? Can you log value returned from `isComplete()` on those resources?
Maybe this is problem with the setup of mocks that you are performing? https://github.com/filipe-costa/jest-timeout/blob/master/src/loader.test.ts

**Reply by filipe-costa** [reply]
Unfortunately, I have poked around with this and isComplete() is not called when I run the tests, I will assume that is failing beforehand and is somehow doing it silently.

Now, I have pulled the repository and I have added this test case to the LoaderBase.tests.ts:

```typescript
    it('should load mocks', (done) =>
    {
        const loader = new Loader();
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore

        loader.add('dataUrlGifMock', fixtureData.dataUrlGif)
            .load(() =>
            {
                // eslint-disable-next-line no-console
                console.log(loader.resources);
                expect(true).to.be.true;
                done();
            });
    });
 ```
 
 I used the following command:
 ```
 npm run test
 ```
 
 And the test hangs when it reaches this test.
 
When I run the test cases using the following command:

```
npm run unit-test:debug
```

The test itself actually passes.
 
Based on this fact and the test cases in LoaderBase.tests.ts, I will have to assume that it might not be possible to mock the Loader API without requiring a running server.

Please, feel free to run the tests as I mentioned above to help confirm my suspicions!

**Reply by domis86** [reply]
> Based on this fact and the test cases in LoaderBase.tests.ts, I will have to assume that it might not be possible to mock the Loader API without requiring a running server.

I see you are mocking the "fetch" - but it seems that PIXI Loader is loading using either XHR ( https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L837 ) or via "Image" in DOM ( https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L728 ) - so probably your "fetch mock" has no effect here.

**Reply by domis86** [reply]
... but i see this test in PIXI code: https://github.com/pixijs/pixijs/blob/dev/packages/loaders/test/Loader.tests.ts#L15 - there a server is actually created and "load" function is tested. Maybe you can do similar like that?

**Reply by filipe-costa** [reply]
Thanks @domis86 , I was afraid that you would say that, I will probably have to implement something similar.

Regarding mocking "fetch" I know for a fact that is not impacting the tests because when I removed it and use PixiLoader by itself with mock JSON files, it still fails.

You have raised a great point regarding via "Image" in DOM, I think this is actually what is happening.

My resource above has the following properties:

```json
loadType: 2,
type: 3
```

The load type stands for image, as you can see in this line: https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L1311

And type also stands for image, as you can see in this line:
https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L1290

This triggers the line that you just mentioned:
https://github.com/pixijs/pixijs/blob/dev/packages/loaders/src/LoaderResource.ts#L728

With this in mind, I wouldn't expect it to fail when running the tests in Jest in the context of the JSDOM environment since Image DOM Web API would exist.

Is it possible that the onComplete signal is not being correctly triggered?

I can for now go with the solution of using the server and I would be more than happy to help investigate this matter further!

**Reply by filipe-costa** [reply]
I have further investigated this matter and these are my findings:

1. JSDOM by default doesn't load subresources, so in the Jest configuration I add to pass the following properties so they can be injected by Jest into the JSDOM constructor:

```
  testEnvironmentOptions: {
    resources: "usable",
  }
```

2. To load the images subresources, JSDOM loads them using the <img> only via the canvas package.

3. I had to enable the esModuleInterop flag in my tsconfig file, not sure if this is related to how Pixijs packages their components or how I am importing things in both my tests and other files or a mix of both. Whichever it is, it has been fun.

When I did the first two the status flag was set to 3, which means that they have been completed/resolved but this was still failing in the test, by changing the flag in point 3, everything worked smoothly.

I would like to say thanks for your help @bigtimebuddy @domis86 !

References:

[JSDOM subresources](https://github.com/jsdom/jsdom#loading-subresources)
[Jest Test EnvironmentOptions](https://jestjs.io/docs/configuration#testenvironmentoptions-object)
[TSConfig esModuleInterop](https://www.typescriptlang.org/tsconfig/#esModuleInterop)

**Comment by filipe-costa** [comment, marked as answer]
I would like to say thanks to @bigtimebuddy and @domis86 for the help.

The answer can be found on the following discussion comment:
https://github.com/pixijs/pixijs/discussions/7862#discussioncomment-1469882

(just updating that this is the answer to the problem in question, so it can be marked as such in the discussions)
